File: frontend/src/domain/astrodynamics/FrameConverter.ts
/**
 * Reference frame transformations for orbital mechanics.
 *
 * Supported frames:
 *   ECI  — Earth-Centered Inertial (J2000/GCRS approximation)
 *   ECEF — Earth-Centered Earth-Fixed (ITRS approximation via GMST)
 *   PQW  — Perifocal (P toward periapsis, Q 90° ahead, W = h-hat)
 *   LVLH — Local Vertical / Local Horizontal (CW rendezvous frame)
 *
 * For production accuracy, ECI↔ECEF should go through the backend
 * which uses astropy's full IAU 2006/2000A precession-nutation model.
 *
 * References:
 *   Vallado §3.1, Curtis §4.4, IERS Conventions 2010
 */

import { computeGMST } from './TimeSystem';
import type { Vec3, OrbitalElements } from './types';
import { MU_EARTH } from './constants';

// ── Vector utilities ────────────────────────────────────────────────────────

export function vecAdd(a: Vec3, b: Vec3): Vec3 {
  return { x: a.x + b.x, y: a.y + b.y, z: a.z + b.z };
}

export function vecScale(v: Vec3, s: number): Vec3 {
  return { x: v.x * s, y: v.y * s, z: v.z * s };
}

export function vecDot(a: Vec3, b: Vec3): number {
  return a.x * b.x + a.y * b.y + a.z * b.z;
}

export function vecCross(a: Vec3, b: Vec3): Vec3 {
  return {
    x: a.y * b.z - a.z * b.y,
    y: a.z * b.x - a.x * b.z,
    z: a.x * b.y - a.y * b.x,
  };
}

export function vecMag(v: Vec3): number {
  return Math.sqrt(v.x * v.x + v.y * v.y + v.z * v.z);
}

export function vecNorm(v: Vec3): Vec3 {
  const m = vecMag(v);
  return m > 0 ? vecScale(v, 1 / m) : { x: 0, y: 0, z: 0 };
}

// ── PQW ↔ ECI ───────────────────────────────────────────────────────────────

/**
 * 3×3 rotation matrix from Perifocal (PQW) to ECI frame.
 * Q = R3(−Ω) · R1(−i) · R3(−ω)
 *
 * Returns only the 6 elements needed when z_pqw = 0 (in-plane orbit):
 * row 0: (r11, r12),  row 1: (r21, r22),  row 2: (r31, r32)
 * Full matrix 3rd column is for out-of-plane component (unused for position on orbit).
 */
export function pqwToEciMatrix(raan: number, inc: number, omega: number): number[] {
  const cosO = Math.cos(raan),  sinO = Math.sin(raan);
  const cosI = Math.cos(inc),   sinI = Math.sin(inc);
  const cosW = Math.cos(omega), sinW = Math.sin(omega);

  return [
    cosO * cosW - sinO * sinW * cosI,   // [0,0] = r11
    -cosO * sinW - sinO * cosW * cosI,  // [0,1] = r12
    sinO * sinI,                         // [0,2] = r13
    sinO * cosW + cosO * sinW * cosI,   // [1,0] = r21
    -sinO * sinW + cosO * cosW * cosI,  // [1,1] = r22
    -cosO * sinI,                        // [1,2] = r23
    sinW * sinI,                         // [2,0] = r31
    cosW * sinI,                         // [2,1] = r32
    cosI,                                // [2,2] = r33
  ];
}

/**
 * Transform a position vector from Perifocal frame to ECI using the 9-element matrix.
 */
export function pqwToEci(r_pqw: Vec3, Q: number[]): Vec3 {
  return {
    x: Q[0] * r_pqw.x + Q[1] * r_pqw.y + Q[2] * r_pqw.z,
    y: Q[3] * r_pqw.x + Q[4] * r_pqw.y + Q[5] * r_pqw.z,
    z: Q[6] * r_pqw.x + Q[7] * r_pqw.y + Q[8] * r_pqw.z,
  };
}

// ── ECI ↔ ECEF ──────────────────────────────────────────────────────────────

/**
 * Rotate ECI position to ECEF using GMST rotation about Z-axis.
 * r_ecef = Rz(θ_GMST) · r_eci
 *
 * Accurate to ~1 km for most applications; for sub-meter accuracy
 * use the backend astropy GCRS→ITRS conversion.
 */
export function eciToEcef(r_eci: Vec3, unixTime: number): Vec3 {
  const theta = computeGMST(unixTime);
  const cos_t = Math.cos(theta);
  const sin_t = Math.sin(theta);
  return {
    x:  cos_t * r_eci.x + sin_t * r_eci.y,
    y: -sin_t * r_eci.x + cos_t * r_eci.y,
    z:  r_eci.z,
  };
}

/** Inverse: ECEF → ECI (rotation by −GMST) */
export function ecefToEci(r_ecef: Vec3, unixTime: number): Vec3 {
  const theta = computeGMST(unixTime);
  const cos_t = Math.cos(theta);
  const sin_t = Math.sin(theta);
  return {
    x: cos_t * r_ecef.x - sin_t * r_ecef.y,
    y: sin_t * r_ecef.x + cos_t * r_ecef.y,
    z: r_ecef.z,
  };
}

// ── ECEF → Geodetic ──────────────────────────────────────────────────────────

/**
 * Convert ECEF position to geodetic latitude, longitude, altitude.
 * Uses Bowring iterative method (WGS-84).
 *
 * @returns { lat: degrees, lon: degrees, alt: km }
 */
export function ecefToGeodetic(r: Vec3): { lat: number; lon: number; alt: number } {
  const a  = 6378.1370;    // WGS-84 equatorial radius, km
  const f  = 1 / 298.257223563;
  const e2 = 2 * f - f * f; // first eccentricity squared

  const lon = Math.atan2(r.y, r.x) * (180 / Math.PI);
  const rho = Math.sqrt(r.x * r.x + r.y * r.y);

  // Bowring iteration
  let lat = Math.atan2(r.z, rho * (1 - e2));
  for (let i = 0; i < 5; i++) {
    const sinLat = Math.sin(lat);
    const N = a / Math.sqrt(1 - e2 * sinLat * sinLat);
    lat = Math.atan2(r.z + e2 * N * sinLat, rho);
  }
  const sinLat = Math.sin(lat);
  const N = a / Math.sqrt(1 - e2 * sinLat * sinLat);
  const alt = rho / Math.cos(lat) - N;

  return { lat: lat * (180 / Math.PI), lon, alt };
}

/** ECI → geodetic (combines ECI→ECEF→geodetic) */
export function eciToGeodetic(r_eci: Vec3, unixTime: number) {
  return ecefToGeodetic(eciToEcef(r_eci, unixTime));
}

// ── Orbital elements utilities ────────────────────────────────────────────────

/**
 * Compute specific angular momentum from orbital elements.
 * h = √(μ·p),  p = a(1−e²)
 */
export function specificAngularMomentum(oe: OrbitalElements): number {
  const p = oe.a * (1 - oe.e * oe.e);
  return Math.sqrt(MU_EARTH * p);
}

File: backend/app/domain/astrodynamics/frame_converter.py
"""
Reference frame conversions using astropy for production accuracy.
Uses IAU 2006/2000A precession-nutation model via GCRS→ITRS.
"""

from datetime import datetime, timezone
import math
import numpy as np

try:
    from astropy.time import Time
    from astropy.coordinates import GCRS, ITRS
    import astropy.units as u
    ASTROPY_AVAILABLE = True
except ImportError:
    ASTROPY_AVAILABLE = False


def eci_to_ecef_astropy(
    r_eci_km: np.ndarray,
    epoch_unix: float,
) -> np.ndarray:
    """
    ECI (GCRS) → ECEF (ITRS) using astropy's full precession-nutation model.
    Accurate to sub-meter level.
    Falls back to GMST rotation if astropy unavailable.
    """
    if not ASTROPY_AVAILABLE:
        return eci_to_ecef_gmst(r_eci_km, epoch_unix)

    t = Time(epoch_unix, format="unix", scale="utc")
    gcrs = GCRS(
        x=r_eci_km[0] * u.km,
        y=r_eci_km[1] * u.km,
        z=r_eci_km[2] * u.km,
        representation_type="cartesian",
        obstime=t,
    )
    itrs = gcrs.transform_to(ITRS(obstime=t))
    xyz  = itrs.cartesian.xyz.to(u.km).value
    return xyz


def eci_to_ecef_gmst(r_eci_km: np.ndarray, epoch_unix: float) -> np.ndarray:
    """
    ECI → ECEF via GMST rotation about Z-axis.
    Simplified (IAU 1982), accurate to ~1 km.
    """
    theta = compute_gmst(epoch_unix)
    c, s  = math.cos(theta), math.sin(theta)
    return np.array([
         c * r_eci_km[0] + s * r_eci_km[1],
        -s * r_eci_km[0] + c * r_eci_km[1],
        r_eci_km[2],
    ])


def compute_gmst(unix_s: float) -> float:
    """Greenwich Mean Sidereal Time, radians. IAU 1982."""
    JD   = unix_s / 86400.0 + 2440587.5
    T    = (JD - 2451545.0) / 36525.0
    theta = (100.4606184
             + 36000.77004 * T
             + 0.000387933 * T**2
             - T**3 / 38710000.0)
    theta = math.fmod(theta, 360.0)
    if theta < 0:
        theta += 360.0
    frac_day = math.fmod(unix_s, 86400.0) / 86400.0
    theta   += 360.98564724 * frac_day
    return math.fmod(theta, 360.0) * math.pi / 180.0


def ecef_to_geodetic(r_ecef_km: np.ndarray) -> dict:
    """ECEF → geodetic latitude, longitude, altitude (WGS-84)."""
    a  = 6378.1370
    f  = 1 / 298.257223563
    e2 = 2 * f - f**2

    x, y, z = r_ecef_km
    lon = math.atan2(y, x)
    rho = math.sqrt(x**2 + y**2)

    lat = math.atan2(z, rho * (1 - e2))
    for _ in range(5):
        sinLat = math.sin(lat)
        N = a / math.sqrt(1 - e2 * sinLat**2)
        lat = math.atan2(z + e2 * N * sinLat, rho)

    sinLat = math.sin(lat)
    N   = a / math.sqrt(1 - e2 * sinLat**2)
    alt = rho / math.cos(lat) - N if abs(math.cos(lat)) > 1e-10 else abs(z) - N * (1 - e2)

    return {
        "lat_deg": lat * 180 / math.pi,
        "lon_deg": lon * 180 / math.pi,
        "alt_km":  alt,
    }
